### ISSUE TITLE
Introduce a filter to control whether an order is partially refunded in `wc_create_refund`

### ISSUE THREAD

**Issue by xristos3490** [woothemes/woocommerce#31416]
### Describe the solution you'd like

Hey team!

While trying to support admin refunds in the WooCommerce Gift Cards extension, we needed to reuse the `wc_create_refund()` function. This function includes a hard check on whether the order is considered partial or fully refunded. See https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-order-functions.php#L613

This check creates an issue for us when an order is considered fully refunded, but there is still some captured gift card left to refund.

It would be great to introduce a filter to control whether an order is partially refunded or not in `wc_create_refund`.

Related issue: https://github.com/woocommerce/woocommerce/issues/31413

### Describe alternatives you've considered

Please visit the P2 project thread for more details: https://wp.me/pdqkbG-6O#comment-169

**Comment by jeffstieler** [comment]
> This check creates an issue for us when an order is considered fully refunded, but there is still some captured gift card left to refund.

If there is some remaining gift card balance to refund, should the order be considered fully refunded?

Would it make more sense to add filters to `WC_Order::get_total()` and `WC_Order::get_total_refunded()` instead?

**Comment by xristos3490** [comment]
Hey @jeffstieler!

> If there is some remaining gift card balance to refund, should the order be considered fully refunded?

When a customer pays for an Order with gift cards, the balance used should be considered a second "payment gateway". With that said, if an Order still has gift card balance captured, it should not be considered fully refunded. 

Some more details. 

An Order might be paid partially with gift cards balance. This transaction needs a separation between the two payment sources. For example, we couldn't preserve a single amount, e.g., `WC_Order::get_total_refunded()`, to summarize the refunded amount in both the "traditional" payment gateway and the refunded gift card balance.

We could think of this process as a reproduction of having multiple payment gateways available to pay for an Order -- which is not supported by WC in general. This assumption makes things a bit more complicated than it seems.

> Would it make more sense to add filters to WC_Order::get_total() and WC_Order::get_total_refunded() instead?

Introducing filters for these amounts opens some new challenges. The most important would be to determine the proper context for filtering those. Given that Gift Cards should have their way of managing the total refunded amount to gift cards (as a separate payment gateway), these filters would play a "hack" role in "cheating" the system to handle an Order as "partially refunded" even if the expression `WC_Order::get_total() = WC_Order::get_total_refunded()` is true.

Moreover, since the "remaining amount to refund" [gets cached in a local variable](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-order-functions.php#L517) early in the execution flow, finding the proper context is even more challenging. Adding context is usually a task of finding the proper actions _before_ and _after_ the target execution place, and with using `did_action` statements, we can be sure that the filter will only run in a specific code segment.

To recap, filtering refund-related amounts isn't the most reasonable way to proceed here because Gift Cards should be viewed as another payment gateway and not a case with aggregated total amounts.

For example, take a look at how the template handles the gift card refunds:
https://d.pr/i/65xdkI

The amounts are separated to clarify that you can only refund to gift cards the amount of the used balance and nothing more.

Another related filter that is needed in the template side of things is https://github.com/woocommerce/woocommerce/pull/31414

I hope this makes sense.
